core/src/main/java/io/questdb/griffin/engine/functions/groupby/StringAggGroupByFunction.java

```diff
@@ -34,100 +34,132 @@
 import io.questdb.griffin.engine.functions.StrFunction;
 import io.questdb.griffin.engine.functions.UnaryFunction;
-import io.questdb.std.Misc;
-import io.questdb.std.ObjList;
-import io.questdb.std.str.DirectUtf16Sink;
+import io.questdb.griffin.engine.groupby.GroupByAllocator;
+import io.questdb.griffin.engine.groupby.GroupByCharSink;
 
 class StringAggGroupByFunction extends StrFunction implements UnaryFunction, GroupByFunction {
-    // Cleared function retains up to INITIAL_SINK_CAPACITY * LIST_CLEAR_THRESHOLD bytes.
-    private static final int INITIAL_SINK_CAPACITY = 512;
-    private static final int LIST_CLEAR_THRESHOLD = 64;
     private final Function arg;
     private final char delimiter;
     private final int functionPosition;
     private final int maxBytes;
-    private final ObjList<DirectUtf16Sink> sinks = new ObjList<>();
-    private int sinkIndex = 0;
-    private int touchedMemorySize;
+
+    private final GroupByCharSink sinkA = new GroupByCharSink();
+    private final GroupByCharSink sinkB = new GroupByCharSink();
+
+    private int totalMemoryUsed;
     private int valueIndex;
 
     public StringAggGroupByFunction(Function arg, int functionPosition, char delimiter, int maxBytes) {
         this.arg = arg;
         this.delimiter = delimiter;
-        this.maxBytes = maxBytes;
         this.functionPosition = functionPosition;
+        this.maxBytes = maxBytes;
     }
 
     @Override
     public void clear() {
-        // Free extra sinks.
-        touchedMemorySize = 0;
-        if (sinks.size() > LIST_CLEAR_THRESHOLD) {
-            for (int i = sinks.size() - 1; i > LIST_CLEAR_THRESHOLD - 1; i--) {
-                Misc.free(sinks.getQuick(i));
-                sinks.remove(i);
-            }
-        }
-        // Reset capacity on the remaining ones.
-        for (int i = 0, n = sinks.size(); i < n; i++) {
-            DirectUtf16Sink sink = sinks.getQuick(i);
-            if (sink != null) {
-                sink.resetCapacity();
-                touchedMemorySize += sink.size();
-            }
-        }
-        sinkIndex = 0;
+        sinkA.of(0);
+        sinkB.of(0);
+        totalMemoryUsed = 0;
     }
 
     @Override
     public void close() {
-        Misc.freeObjListAndClear(sinks);
+        sinkA.clear();
+        sinkB.clear();
     }
 
     @Override
     public void computeFirst(MapValue mapValue, Record record, long rowId) {
-        final DirectUtf16Sink sink;
-        if (sinks.size() <= sinkIndex) {
-            sinks.extendAndSet(sinkIndex, sink = new DirectUtf16Sink(INITIAL_SINK_CAPACITY));
-        } else {
-            sink = sinks.getQuick(sinkIndex);
-            sink.clear();
-        }
-
         final CharSequence str = arg.getStrA(record);
+
         if (str != null) {
-            sink.put(str);
+            //Non-null value: initialize a new sink
+            sinkA.of(0);
+            sinkA.put(str);
+
+            totalMemoryUsed += sinkA.length() * 2;  // UTF-16: 2 bytes per char
```

Comment:
You could consider using the allocator to check how much memory the current thread (or even all threads) has allocated.